### ISSUE TITLE
In-Place operations on items inside a Union of any MutableSequence produce an unexpected error.

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#4700]
**Describe the bug**
Pyright produces an unexpected error when applying an in-place operation on an item in a sequence, in situations where the type is evaluated to be a union of multiple sequence types. (e.g. `list[int] | list[str]`)

**To Reproduce**
See the code below, pyright will output the following errors.

```
  test.py:7:9 - error: Argument of type "int" cannot be assigned to parameter "index" of type "slice" in function "__setitem__"
    "int" is incompatible with "slice" (reportGeneralTypeIssues)
  test.py:7:9 - error: Argument of type "int | float" cannot be assigned to parameter "value" of type "Iterable[int]" in function "__setitem__"
    Type "int | float" cannot be assigned to type "Iterable[int]"
      "float" is incompatible with protocol "Iterable[int]"
        "__iter__" is not present (reportGeneralTypeIssues)
```

**Expected behavior**
No error, code shown below is valid.

**Screenshots or Code**

```python
from __future__ import annotations
from collections.abc import MutableSequence


def foo(data: MutableSequence[int] | MutableSequence[float]):
    for i, _ in enumerate(data):
        data[i] += 3
```

**VS Code extension or command-line**
pyright 1.1.295

**Comment by erictraut** [comment]
The current behavior is as designed, so I don't consider this a bug. Mypy generates the same error in this case.

The code in question is logically equivalent to:
```python
def foo(data: MutableSequence[int] | MutableSequence[float]):
    x = data[0] + 3
    reveal_type(x) # int | float
    data[0] = x
```

The type of `x` in this case is `int | float`, and you cannot assign this value back to `data[0]`.
